Add primal_internal.h

amontoison opened the pull request:
`primal_internal.h` contains all headers for the C interface.
It includes `prima.h` such that we can only the export routines in `primal.h` and the unexported ones in `primal_internal.h`.

amontoison commented:
> Thank you @amontoison . The errors complain that some function is deprecated and suggest a replacement. It seems easy to fix, right? Thanks.

For the issue with deprecated routine, we can use the preprocessed macro `WIN32_` to check if we are on Windows.

For the alignment error, I don't know what we should do.

zaikunzhang commented:
Hi @amontoison ,

> For the issue with deprecated routine, we can use the preprocessed macro _WIN32 to check if we are on Windows.


```c
// Thread-safe version of localtime
#ifdef _WIN32
#define localtime_safe(a, b) localtime_s(a, b)
#else
#define localtime_safe(a, b) localtime_r(b, a)
#endif
```


Why do we need to differentiate `_WIN32` from others? Could we just use `local_time_s` in both cases (I note that `local_time_r` is available only with C23, which is too recent). In addition, why `a` and `b` are reversed in the two cases? 

Thanks. 

Zaikun

zaikunzhang commented:
https://github.com/amontoison/prima/blob/c_headers_internal/c/tests/stress.c#L10
```c
// Make PRIMA available
#include "prima/prima_internal.h"
```


Shouldn't we use `prima.h` instead of `prima_internal.h` here?

amontoison commented:
> https://github.com/amontoison/prima/blob/c_headers_internal/c/tests/stress.c#L10
> 
> ```c
> // Make PRIMA available
> #include "prima/prima_internal.h"
> ```
> 
> Shouldn't we use `prima.h` instead of `prima_internal.h` here?

You need the header of `get_random_sed` in `stress.c` nd I don't think that you want to export it.

amontoison commented:
> Hi @amontoison ,
> 
> > For the issue with deprecated routine, we can use the preprocessed macro _WIN32 to check if we are on Windows.
> 
> ```c
> // Thread-safe version of localtime
> #ifdef _WIN32
> #define localtime_safe(a, b) localtime_s(a, b)
> #else
> #define localtime_safe(a, b) localtime_r(b, a)
> #endif
> ```
> 
> Why do we need to differentiate `_WIN32` from others? Could we just use `localtime_s` in both cases (I note that `localtime_r` is available only with C23, which is too recent)?
> 
> In addition, why are `a` and `b` reversed in the two cases?
> 
> Thanks.
> 
> Zaikun

Hi @zaikunzhang, I checked online and it seems that `localtime_s` is only available on Windows.
If `localtime_s` is too recent, we should keep `localtime` on the other platform. We don’t have warnings is a thread-safe version is not used.

zaikunzhang commented:
Hi @amontoison ,

I wanted to try fixing other warnings but did not know how to do it cleanly on top of your PR, so I created a new one. I hope you do not mind. See https://github.com/libprima/prima/pull/196 .

> Hi @zaikunzhang, I checked online and it seems that localtime_s is only available on Windows.

I checked and I agree. 

> You need the header of get_random_sed in stress.c and I don't think that you want to export it.

I changed its definition to `static unsigned int ...` and removed it from `prima_internal.h`. 

Could you review https://github.com/libprima/prima/pull/196 ? Thank you. 

Zaikun

amontoison commented:
> Hi @amontoison ,
> 
> I wanted to try fixing other warnings but did not know how to do it cleanly on top of your PR, so I created a new one. I hope you do not mind. See #196 .
> 
> > Hi @zaikunzhang, I checked online and it seems that localtime_s is only available on Windows.
> 
> I checked and I agree.
> 
> > You need the header of get_random_sed in stress.c and I don't think that you want to export it.
> 
> I changed its definition to `static unsigned int ...` and removed it from `prima_internal.h`.
> 
> Could you review #196 ? Thank you.
> 
> Zaikun

@zaikunzhang 
Do you still want to keep `localtime_safe(a, b) localtime_r(b, a)` for platforms that are not Windows?
You can use `localtime_safe(a, b) localtime(a, b)` if `C23` is a constraint.

zaikunzhang commented:
> > 
> 
> @zaikunzhang Do you still want to keep `localtime_safe(a, b) localtime_r(b, a)` for platforms that are not Windows? You can use `localtime_safe(a, b) localtime(a, b)` if `C23` is a constraint.

Hi @amontoison ,

It turns out that `localtime_s` is not available on macOS. Since it is only for test, there is no big problem to use `C23`. Thank you.
